combat-trainer.lic

```diff
@@ -295,73 +298,76 @@ class LootProcess
-      check_skinning(DRRoom.dead_npcs.first, game_state)
-    end
-    unless healing
-      while 'and get ready to search it' == bput('loot', 'You search', 'I could not find what you were referring to', 'and get ready to search it')
-        pause
-        waitrt?
+
+      unless game_state.prepare_consume
+        if @ritual_type == 'harvest'
+          do_ritual(DRRoom.dead_npcs.first, 'preserve', game_state)
+          do_ritual(DRRoom.dead_npcs.first, @ritual_type, game_state)
+        elsif @ritual_type
+          do_ritual(DRRoom.dead_npcs.first, @ritual_type, game_state)
+        end
       end
-      @last_ritual = nil
     end
-    @loot_timer = Time.now
   end
 
-  def check_thanatology(mob_noun, ritual, game_state)
-    return false if game_state.construct?(mob_noun)
+  def do_ritual(mob_noun, ritual, game_state)
+    return unless DRStats.necromancer?
+    return if game_state.construct?(mob_noun)
 
     perform_message = "perform #{ritual} on #{mob_noun}"
-    case bput(perform_message, 'Rituals do not work upon constructs', 'Rituals do not work upon the undead', 'This ritual may only be performed on a corpse', 'precise motions with your ritual knife', 'corpse to make one long', 'carefully position the corpse', 'corpse and make a few quick precise cuts', 'roundtime')
-    when 'corpse to make one long', 'carefully position the corpse', 'corpse and make a few quick precise cuts', 'Your cuts follow the vital lines of the corpse'
+    case bput(perform_message, 'Rituals do not work upon constructs', 'Rituals do not work upon the undead', 'This ritual may only be performed on a corpse', @rituals['preserve'], @rituals['dissect'], @rituals['harvest'], @rituals['consume'], 'roundtime')
+    when @rituals['preserve'], @rituals['dissect']
       @last_ritual = ritual
-      return true
-    when 'precise motions with your ritual knife'
+      return
+    when @rituals['consume']
+      @last_ritual = ritual
+      game_state.prepare_consume = true if @heal
+      return
+    when @rituals['harvest']
       @last_ritual = ritual
       waitrt?
       fput 'drop material'
-      return true
+      return
     when 'Rituals do not work upon constructs'
       game_state.construct(mob_noun)
-      return false
-    when 'Rituals do not work upon the undead', 'This ritual may only be performed on a corpse'
-      return false
+      return
     end
-    @last_ritual = ritual
-    waitrt?
-    pause 0.25
-    waitrt?
-    snap = [checkleft, checkright]
-    dispose_trash(checkleft) if snap.first != checkleft
-    dispose_trash(checkright) if snap.last != checkright
-    true
+  end
+
+  def dispose_body(game_state)
+    return unless @loot_bodies
+    if DRRoom.dead_npcs.empty?
+      @last_ritual = nil
+      return
+    end
+    return if Time.now - @loot_timer < @loot_delay
+
+    game_state.mob_died = true
+    check_rituals(game_state)
+    if @skin && !%w(consume harvest dissect).include?(@last_ritual)
```

Comment:
I'm trying to think about how this would work with a list. Is there a sensible way to define do_ritual to work on both a string or a list as input? Most situations will call for a single ritual, whatever the user has defined to be done, so that would be passed as a string. However some rituals require a preserve ritual to be done first, so that could be passed as an array.

A second thought: do ritual can just check if the ritual arg is 'harvest', 'arise', or 'consume' (the rituals that make use of preserve) then apply preserve on its own